src/reduce.zig

```diff
@@ -182,51 +182,96 @@ pub fn main(gpa: Allocator, arena: Allocator, args: []const []const u8) !void {
             rendered.clearRetainingCapacity();
             try tree.renderToArrayList(&rendered, fixups);
 
-            // The transformations we applied may have resulted in unused locals,
+            // The transformations we applied may have resulted in unused locals and similar artifacts,
             // in which case we would like to add the respective discards.
-            {
+            var found_uncorrected_errors = false;
+            var validation_run = false;
+            while (true) {
                 try astgen_input.resize(rendered.items.len);
                 @memcpy(astgen_input.items, rendered.items);
                 try astgen_input.append(0);
                 const source_with_null = astgen_input.items[0 .. astgen_input.items.len - 1 :0];
                 var astgen_tree = try Ast.parse(gpa, source_with_null, .zig);
                 defer astgen_tree.deinit(gpa);
-                if (astgen_tree.errors.len != 0) {
+                if (astgen_tree.errors.len != 0) blk: {
+                    if (validation_run) {
+                        found_uncorrected_errors = true;
+                        break;
+                    }
+                    var found_other_astgen_errors = false;
+                    for (astgen_tree.errors) |err| {
+                        if (err.tag == .expected_statement and astgen_tree.tokens.items(.tag)[err.token] == .semicolon) {
+                            const byte_offset = astgen_tree.tokens.items(.start)[err.token];
+                            rendered.items[byte_offset] = ' ';
+                            source_with_null[byte_offset] = ' ';
+                        } else found_other_astgen_errors = true;
+                        // std.debug.print("astgen error: {any}\n", .{err});
+                    }
+                    if (!found_other_astgen_errors) {
+                        astgen_tree.deinit(gpa);
+                        astgen_tree = try Ast.parse(gpa, source_with_null, .zig);
+                        if (astgen_tree.errors.len == 0)
+                            break :blk;
+                    }
+                    std.fs.cwd().writeFile(root_source_file_path, source_with_null) catch
+                        std.debug.print("failed to write source file prior panic\n", .{});
```

Comment:
Do you really need to introduce the complexity of having the render code possibly emit invalid syntax? Is this for the leftover semicolon from branch reduction?

I think I would rather remove that semicolon from the language altogether... but in the meantime, render should be able to omit the semicolon as appropriate and then all this logic can be omitted.

Reply:
> Do you really need to introduce the complexity of having the render code possibly emit invalid syntax? Is this for the leftover semicolon from branch reduction?

Yes, that's the reason. I had looked into not emitting in the first place but since the semicolon lives outside the branch there was no simple fix.
I can look into it again but I don't know yet when I'll have the time.

Reply:
With 70c58ae the semicolon will no longer be emitted. Whether this reduced the complexity I don't know.

Brief explanation on this change:
- Node replacemnets are applied up front so that if/for/while expressions know if their branches are blocks or not. This is necessary since they are now responsible for deciding on whether or not to emit a semicolon.
- Due to this, a copy of the original nodes is maintained. This is necessary so that 1. the original code can be restored between iterations, and 2. because a mutated tree causes `tree.lastToken(node)` to report incorrect results. (It would e.g. emit an `else` instead of a `}` if branch reduction occured within the block getting finished.)
- A new `Space` tag `.smart` is introduced to signal to a (if/for/while) node that it is responsible for deciding on the space it propagates. This signaling is necessary because there's way too many edge-cases without.